src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.BufferToTagTree.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+internal partial class AbstractAsynchronousTaggerProvider<TTag>
+{
+    private readonly struct BufferToTagTree(ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> map)
+    {
+        public static readonly BufferToTagTree Empty = new(ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>>.Empty);
+
+        public readonly ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> Map = map;
+
+        public bool IsDefault => Map is null;
+
+        public TagSpanIntervalTree<TTag> this[ITextBuffer buffer] => Map[buffer];
+
+        internal static BufferToTagTree InterlockedExchange(ref BufferToTagTree location, BufferToTagTree value)
+            => new(Interlocked.Exchange(ref Unsafe.AsRef(in location.Map), value.Map));
```

Comment:
Is this just more sane to have the struct use a mutable field and then you don't need this directly? Or can dispense of the Unsafe.AsRef? I normally say that a mutable struct is terrible, but this is already effectively mutable so maybe it doesn't really matter.

Reply:
removed this all.  back to ImmutableDictionaries.